metadata_templates.py

```diff
@@ -64,10 +77,82 @@ def metadata_templates_collection_detach(rule_args, callback, rei):
     result = mt_detach(callback, logical_path, 'collection', schema, type)
 
 
+def metadata_templates_data_object_gather(rule_args, callback, rei):
+    # Export/Collapse/Rasterize/Gather/Dump (logical_path, recursive, schemas_string)
+    # - Find all associated schemas
+    # - Recursive gathers schemas on all parents up to root
+
+    logical_path = rule_args[0]
+    recursive = rule_args[1]
+
+    # split logical path
+    collection_name, data_name = logical_path.rsplit("/", 1)
+
+    # get all schema locations attached to this data object
+    schemas = []
+    for schema, thetype in Query(callback,
+                        "META_DATA_ATTR_VALUE, META_DATA_ATTR_UNITS",
+                        "COLL_NAME = '{}' and DATA_NAME = '{}' and META_DATA_ATTR_NAME = '{}'".format(collection_name, data_name, MT_NAMESPACE)):
+#        callback.writeLine('serverLog','{} {}'.format(thetype, schema))
+        if thetype == 'url':
+            try:
+                # get the schema content from the location
+                r = requests.get(schema)
+                # convert to json object
+                j = json.loads(r.content)
+                # add to schemas array
+                schemas.append(j)
+            except Exception as e:
+                callback.writeLine('serverLog', '{}'.format(type(e)))
+        elif thetype == 'irods':
+            # get length of iRODS full logical path
+            ret = callback.msiObjStat(schema, irods_types.RodsObjStat())
+            objstat = ret['arguments'][1]
+            # get schema contents from iRODS full logical path
+            ret = callback.msiDataObjOpen(schema, 0)
+            fd = ret['arguments'][1] # iRODS file descriptor
+            ret = callback.msiDataObjRead(fd, objstat.objSize, irods_types.BytesBuf())
+            buf = ret['arguments'][2] # bytesbuf
```

Comment:
Hmm, this will allocate / read as much memory as it can based on the given object size.

You may want to introduce an option that allows admins to control the max number of bytes that can be allocated / read.

Reply:
have been talking about this, @d-w-moore may take a swing at a blocksize read after we merge.

have already introduced a commit to remove the stat() roundtrip, and just use a big number until we have a read-loop.

Reply:
created https://github.com/irods-contrib/irods_working_group_metadata_templates/issues/40